Add utility downloader

muschellij2 opened the pull request:
I added functions to download the utilities so that if `ukbmd5` and such would be in your PATH, then `Sys.which` finds it, it will use that.  Otherwise, it will download that binary file and run `Sys.chmod` so that it can run it as an executable.  By default, it downloads them to the temporary directory to not violate CRAN policy.  

For example:
```r
file = "ukbXXXX.enc"
checksum = "XXXasdf4XXXX"
out = ukb_md5(file)
stopifnot(out == checksum)

out = ukb_unpack(file, key = keyfile)
```

muschellij2 committed:
- change .update_tab_path to be relative if path = "." for ukb_df

muschellij2 committed:
added desc

muschellij2 committed:
merged with master

muschellij2 committed:
Merging with the upstream branch
Merge remote-tracking branch 'upstream/master'

muschellij2 committed:
merged with PR

muschellij2 committed:
fixing typo in the Rd file

muschellij2 committed:
added utilities

muschellij2 committed:
udated the rd

muschellij2 committed:
updated the utilities

muschellij2 committed:
updated the utilities

muschellij2 committed:
updated news

muschellij2 committed:
aded fetch

muschellij2 committed:
updated key

muschellij2 committed:
updated the normalizePath

muschellij2 committed:
added temporary argument to get around permissions failures

muschellij2 committed:
needed to pass r_file through

muschellij2 committed:
adde encoding file

muschellij2 committed:
moved the fiel to later to keep code working

muschellij2 committed:
updated fileset

muschellij2 commented:
Any thoughts on merging this?

Also - do you think adding in the option to `ukb_df` to put in a CSV that has the withdraw eids?

muschellij2 committed:
updated ukb_df to have withdraw_file added

muschellij2 committed:
added warning in case eid is not available

kenhanscombe commented:
Hey, sorry for the delay just looking at this now. Good idea. Mainly updating documentation and style. Will push commits to your branch shortly.

kenhanscombe committed:
Code style and naming updates

Edited code style to match ukbtools package, and ukbtools and UKB
naming conventions. Also added documentation for each function.

kenhanscombe commented:
Could you separate "utilities" and "withdrawals" additions, preferrably to separate feature branches?

We may be able to incorporate the withdrawals additions before the utilities – I could not get noah to work on macOS 10.14.6 mojave. More generally I'm concerned that noah is archived and therefore no longer supported, and my problems with noah are in fact an unresolved issue [#43](https://github.com/linux-noah/noah/issues/43).

I think I'm inclined to take the addition of utilies in a different direction: I've created a [docker image](https://hub.docker.com/repository/docker/onekenken/alpine-ukb-progs) which contains all the md5 checksum'd programs and utilities. It's seems far easier to bind mount to a local directory and `ukbunpack`, `ukbconv`, etc. using the linux container (regardless of local operating system). Yes, it will require users to install docker but that's straightforward and supported. Will push this functionality to a feature branch shortly.

Depending how well that works, we may still use your "direct" approach for linux and windows users.

muschellij2 commented:
I don't have the time to separate them out right now. I think Docker is
fine, and while I agree it's a straightforward install, I think 1) it's a
heavy lift for most users, 2) if you're working with this data it's likely
on a cluster (aka linux) and probably can't use docker (but probably can
use singularity), but I agree that you lose out on macOS users (but UK
Biobank made that decision really for you).

If you go down the Docker road, I think
https://rdrr.io/cran/xaringan/man/decktape.html is a good example of it
used in a package.

Best,
John


On Tue, Sep 1, 2020 at 8:19 AM Ken Hanscombe <notifications@github.com>
wrote:

> Could you separate "utilities" and "withdrawals" additions, preferrably to
> separate feature branches?
>
> We may be able to incorporate the withdrawals additions before the
> utilities – I could not get noah to work on macOS 10.14.6 mojave. More
> generally I'm concerned that noah is archived and therefore no longer
> supported, and my problems with noah are in fact an unresolved issue #43
> <https://github.com/linux-noah/noah/issues/43>.
>
> I think I'm inclined to take the addition of utilies in a different
> direction: I've created a docker image
> <https://hub.docker.com/repository/docker/onekenken/alpine-ukb-progs>
> which contains all the md5 checksum'd programs and utilities. It's seems
> far easier to bind mount to a local directory and ukbunpack, ukbconv,
> etc. using the linux container (regardless of local operating system). Yes,
> it will require users to install docker but that's straightforward and
> supported. Will push this functionality to a feature branch shortly.
>
> Depending how well that works, we may still use your "direct" approach for
> linux and windows users.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kenhanscombe/ukbtools/pull/29#issuecomment-684811049>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIGPLRYURNVOEWXYAEZGULSDTRGVANCNFSM4OFBDVDA>
> .
>

muschellij2 commented:
I guess my overall recommendation is to do an OS switch and use Docker for
OSX users.  I don't know what you'd want to do for Solaris and other OSes
that CRAN supports, however.
Best,
John


On Tue, Sep 1, 2020 at 8:53 AM John Muschelli <muschellij2@gmail.com> wrote:

> I don't have the time to separate them out right now. I think Docker is
> fine, and while I agree it's a straightforward install, I think 1) it's a
> heavy lift for most users, 2) if you're working with this data it's likely
> on a cluster (aka linux) and probably can't use docker (but probably can
> use singularity), but I agree that you lose out on macOS users (but UK
> Biobank made that decision really for you).
>
> If you go down the Docker road, I think
> https://rdrr.io/cran/xaringan/man/decktape.html is a good example of it
> used in a package.
>
> Best,
> John
>
>
> On Tue, Sep 1, 2020 at 8:19 AM Ken Hanscombe <notifications@github.com>
> wrote:
>
>> Could you separate "utilities" and "withdrawals" additions, preferrably
>> to separate feature branches?
>>
>> We may be able to incorporate the withdrawals additions before the
>> utilities – I could not get noah to work on macOS 10.14.6 mojave. More
>> generally I'm concerned that noah is archived and therefore no longer
>> supported, and my problems with noah are in fact an unresolved issue #43
>> <https://github.com/linux-noah/noah/issues/43>.
>>
>> I think I'm inclined to take the addition of utilies in a different
>> direction: I've created a docker image
>> <https://hub.docker.com/repository/docker/onekenken/alpine-ukb-progs>
>> which contains all the md5 checksum'd programs and utilities. It's seems
>> far easier to bind mount to a local directory and ukbunpack, ukbconv,
>> etc. using the linux container (regardless of local operating system). Yes,
>> it will require users to install docker but that's straightforward and
>> supported. Will push this functionality to a feature branch shortly.
>>
>> Depending how well that works, we may still use your "direct" approach
>> for linux and windows users.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/kenhanscombe/ukbtools/pull/29#issuecomment-684811049>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAIGPLRYURNVOEWXYAEZGULSDTRGVANCNFSM4OFBDVDA>
>> .
>>
>

kenhanscombe commented:
Thanks, John.

I think you're right; we'll have to use an OS switch of sorts with docker for mac OS – singularity will work for linux but not for mac (beta version exists but future seems uncertain). Not sure about other OSs. I'm pretty sure I can hide the docker interactions in the backend so that users are only exposed to (pretty much) the list of functions you wrote.

Considering above, I'll work on incorporating the OS switch option (with docker for mac OS, and any other OSs that support docker) into your PR. Will push when I have something substantive.

muschellij2 committed:
Merge remote-tracking branch 'upstream/master' into master

muschellij2 committed:
Merge branch 'master' of https://github.com/muschellij2/ukbtools into master

muschellij2 committed:
updated the withdrawal_file argument
